contracts/clearing/TradingApp.sol

```diff
@@ -92,46 +61,105 @@ contract TradingApp is IForceMoveApp {
 				candidateData,
-				(ITradingStructs.Order)
+				(ITradingTypes.Order)
 			);
 			return (true, '');
 		}
 
 		// orderResponse
 		// NOTE: used just to check the data structure validity
-		ITradingStructs.OrderResponse memory _prevOrderResponse = abi.decode(
+		ITradingTypes.OrderResponse memory _prevOrderResponse = abi.decode(
 			proof0.appData,
-			(ITradingStructs.OrderResponse)
+			(ITradingTypes.OrderResponse)
 		);
 
-		ITradingStructs.Order memory order = abi.decode(proof1.appData, (ITradingStructs.Order));
-		ITradingStructs.OrderResponse memory orderResponse = abi.decode(
+		ITradingTypes.Order memory order = abi.decode(proof1.appData, (ITradingTypes.Order));
+		ITradingTypes.OrderResponse memory orderResponse = abi.decode(
 			candidateData,
-			(ITradingStructs.OrderResponse)
+			(ITradingTypes.OrderResponse)
 		);
-		if (orderResponse.responseType == ITradingStructs.OrderResponseType.ACCEPT) {
+		if (orderResponse.responseType == ITradingTypes.OrderResponseType.ACCEPT) {
 			require(
 				orderResponse.orderID == order.orderID,
 				'orderResponse.orderID != order.orderID, candidate is orderResponse'
 			);
 		}
 		return (true, '');
 	}
+
+	function verifyProofForSettlement(
+		ITradingTypes.Settlement memory settlement,
+		RecoveredVariablePart[] calldata proof
+	) internal pure {
+		if (proof.length == 1) {
+			VariablePart memory proof = proof[0].variablePart;
+			ITradingTypes.Order memory order = abi.decode(proof.appData, (ITradingTypes.Order));
+			return;
+		}
+
+		bytes32[] memory orderIDs = new bytes32[](proof.length);
+		for (uint256 i = 0; i < proof.length - 1; i++) {
+			VariablePart memory currProof = proof[i].variablePart;
+			VariablePart memory nextProof = proof[i + 1].variablePart;
+
+			require(currProof.turnNum + 1 == nextProof.turnNum, 'turns are not consecutive');
+
+			// Verify validity of orders and responses
+			if (i % 2 == 0) {
+				// If current proof contains an order,
+				// then the next one must contain a response
+				// with the same order ID
+				ITradingTypes.Order memory order = abi.decode(
+					currProof.appData,
+					(ITradingTypes.Order)
+				);
+				ITradingTypes.OrderResponse memory orderResponse = abi.decode(
+					nextProof.appData,
+					(ITradingTypes.OrderResponse)
+				);
+				require(
+					orderResponse.orderID == order.orderID,
+					'orderResponse.orderID != order.orderID'
+				);
+				orderIDs[i] = order.orderID;
+			} else {
+				// If current proof contains a response,
+				// then the next one must be an order
+				// with a different order ID, since they are not related
+				ITradingTypes.OrderResponse memory orderResponse = abi.decode(
+					currProof.appData,
+					(ITradingTypes.OrderResponse)
+				);
+				ITradingTypes.Order memory order = abi.decode(
+					nextProof.appData,
+					(ITradingTypes.Order)
+				);
+				require(
+					order.orderID != orderResponse.orderID,
+					'order.orderID == orderResponse.orderID'
+				);
+				orderIDs[i] = order.orderID;
+			}
```

Comment:
I think we can omit this part and increment the iterator by 2, skipping the OrderResponse case. This check should be done by both parties offline before signing the Settlement.